src/XRP/xrp-transaction.ts

```diff
@@ -0,0 +1,110 @@
+import { Transaction } from '../generated/web/org/xrpl/rpc/v1/transaction_pb'
+import RippledFlags from '../rippled-flags'
+import XRPSigner from './xrp-signer'
+import XRPTransactionType from './xrp-transaction-type'
+import XRPPayment from './xrp-payment'
+import XRPMemo from './xrp-memo'
+
+/*
+ * A transaction on the XRP Ledger.
+ *
+ * @see: https://xrpl.org/transaction-formats.html
+ */
+// TODO(amiecorso): Modify this object to use X-Address format.
+export default class XRPTransaction {
+  public static from(transaction: Transaction): XRPTransaction | undefined {
+    const account = transaction
+      .getAccount()
+      ?.getValue()
+      ?.getAddress()
+
+    const fee = transaction.getFee()?.getDrops()
+
+    const sequence = transaction.getSequence()?.getValue()
+
+    const signingPublicKey = transaction.getSigningPublicKey()?.getValue_asU8()
+
+    const transactionSignature = transaction
+      .getTransactionSignature()
+      ?.getValue_asU8()
+
+    const accountTransactionID = transaction
+      .getAccountTransactionId()
+      ?.getValue_asU8()
+
+    const flags = transaction.getFlags()?.getValue()
+
+    const lastLedgerSequence = transaction.getLastLedgerSequence()?.getValue()
+
+    let memos
+    if (transaction.getMemosList().length > 0) {
+      memos = transaction.getMemosList().map((memo) => XRPMemo.from(memo))
+    } else {
+      memos = undefined
+    }
+
+    let signers
+    if (transaction.getSignersList().length > 0) {
+      signers = transaction
+        .getSignersList()
+        .map((signer) => XRPSigner.from(signer))
+    } else {
+      signers = undefined
+    }
+
+    const sourceTag = transaction.getSourceTag()?.getValue()
+
+    let paymentFields
+    let type
+    switch (transaction.getTransactionDataCase()) {
+      case Transaction.TransactionDataCase.PAYMENT: {
+        const payment = transaction.getPayment()
+        if (payment) {
+          paymentFields = XRPPayment.from(payment)
+          type = Transaction.TransactionDataCase.PAYMENT
+          if (!paymentFields) {
+            return undefined
+          }
+        } else {
+          return undefined
```

Comment:
Nice of an either monad (undefined | tx)! Wondering if you need any more information than `undefined` at the layer above? Or is it sufficient information for all of these to return `undefined`?

Reply:
it's sufficient for these to return undefined.  the .from() initializers in all of these related classes should return undefined given certain problems with conversation from the protobufs.  However I did manage to shorten this logic quite a bit

Reply:
(I've learned some stuff from PRs 1 - 8 and had forgotten to incorporate here.)

Reply:
Great!